File: rekordbox/packages/core/src/util/naming.ts
const JUNK_PATTERNS: RegExp[] = [
  /\bofficial\b/gi,
  /\bvideo\b/gi,
  /\blyric(s)?\b/gi,
  /\bvisualizer\b/gi,
  /\bhd\b/gi,
  /\b4k\b/gi
];

export function guessArtistTitle(rawTitle: string, uploader?: string): { artist: string; title: string } {
  let title = rawTitle;
  title = title.replace(/\[[^\]]*]/g, " ");
  title = title.replace(/\([^)]*\bofficial\b[^)]*\)/gi, " ");
  for (const re of JUNK_PATTERNS) title = title.replace(re, " ");
  title = title.replace(/\s+/g, " ").trim();

  const parts = title.split(" - ").map((p) => p.trim()).filter(Boolean);
  if (parts.length >= 2) {
    return { artist: parts[0]!, title: parts.slice(1).join(" - ") };
  }

  return { artist: (uploader ?? "Unknown Artist").trim(), title };
}

export function makeRekordboxFilename(input: { artist: string; title: string; bpm?: number | null; key?: string | null; ext: string }): string {
  const bpm = input.bpm ? Math.round(input.bpm) : null;
  const suffix = bpm || input.key ? ` [${[bpm ? String(bpm) : null, input.key ?? null].filter(Boolean).join(" ")}]` : "";
  return sanitizeFileComponent(`${input.artist} - ${input.title}${suffix}`) + input.ext;
}

export function sanitizeFileComponent(value: string): string {
  // Cross-platform safe-ish: avoid reserved chars and trailing dots/spaces.
  const cleaned = value
    .replace(/[\\/:*?"<>|]/g, " ")
    .replace(/\s+/g, " ")
    .trim()
    .replace(/[. ]+$/g, "");
  return cleaned.length ? cleaned : "Untitled";
}

File: packages/api/dropcrate/services/naming.py
"""Port of packages/core/src/util/naming.ts — Filename sanitization."""

from __future__ import annotations

import re


def sanitize_file_component(value: str) -> str:
    cleaned = re.sub(r'[\\/:*?"<>|]', " ", value)
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    cleaned = re.sub(r"[. ]+$", "", cleaned)
    return cleaned if cleaned else "Untitled"


def make_rekordbox_filename(
    artist: str, title: str, ext: str, bpm: int | None = None, key: str | None = None
) -> str:
    parts = [str(bpm) if bpm else None, key]
    suffix_parts = [p for p in parts if p]
    suffix = f" [{' '.join(suffix_parts)}]" if suffix_parts else ""
    return sanitize_file_component(f"{artist} - {title}{suffix}") + ext
